Mathlib/Computability/Primrec.lean

```diff
@@ -0,0 +1,1618 @@
+#align primrec.of_equiv_iff Primrec.ofEquiv_iff
+
+theorem ofEquiv_symm_iff {β} (e : β ≃ α) {f : σ → α} :
+    haveI := Primcodable.ofEquiv α e
+    (Primrec fun a => e.symm (f a)) ↔ Primrec f :=
+  letI := Primcodable.ofEquiv α e
+  ⟨fun h => (ofEquiv.comp h).of_eq fun a => by simp, ofEquiv_symm.comp⟩
+#align primrec.of_equiv_symm_iff Primrec.ofEquiv_symm_iff
+
+end Primrec
+
+namespace Primcodable
+
+open Nat.Primrec
+
+instance prod {α β} [Primcodable α] [Primcodable β] : Primcodable (α × β) :=
+  ⟨((cases zero ((cases zero .succ).comp (pair right ((@Primcodable.prim β).comp left)))).comp
+          (pair right ((@Primcodable.prim α).comp left))).of_eq
+      fun n => by
+      simp [Nat.unpaired]
+      cases @decode α _ n.unpair.1; · simp
+      cases @ decode β _ n.unpair.2 <;> simp⟩
+#align primcodable.prod Primcodable.prod
+
+end Primcodable
+
+namespace Primrec
+
+variable {α : Type _} {σ : Type _} [Primcodable α] [Primcodable σ]
+
+open Nat.Primrec
+
+theorem fst {α β} [Primcodable α] [Primcodable β] : Primrec (@Prod.fst α β) :=
+  ((cases zero
+            ((cases zero (Nat.Primrec.succ.comp left)).comp
+              (pair right ((@Primcodable.prim β).comp left)))).comp
+        (pair right ((@Primcodable.prim α).comp left))).of_eq
+    fun n => by
+    simp
+    cases @decode α _ n.unpair.1 <;> simp
+    cases @decode β _ n.unpair.2 <;> simp
+#align primrec.fst Primrec.fst
+
+theorem snd {α β} [Primcodable α] [Primcodable β] : Primrec (@Prod.snd α β) :=
+  ((cases zero
+            ((cases zero (Nat.Primrec.succ.comp right)).comp
+              (pair right ((@Primcodable.prim β).comp left)))).comp
+        (pair right ((@Primcodable.prim α).comp left))).of_eq
+    fun n => by
+    simp
+    cases @decode α _ n.unpair.1 <;> simp
+    cases @decode β _ n.unpair.2 <;> simp
+#align primrec.snd Primrec.snd
+
+theorem pair {α β γ} [Primcodable α] [Primcodable β] [Primcodable γ] {f : α → β} {g : α → γ}
+    (hf : Primrec f) (hg : Primrec g) : Primrec fun a => (f a, g a) :=
+  ((cases1 0
+            (Nat.Primrec.succ.comp <|
+              .pair (Nat.Primrec.pred.comp hf) (Nat.Primrec.pred.comp hg))).comp
+        (@Primcodable.prim α _)).of_eq
+    fun n => by cases @decode α _ n <;> simp [encodek]
+#align primrec.pair Primrec.pair
+
+theorem unpair : Primrec Nat.unpair :=
+  (pair (nat_iff.2 .left) (nat_iff.2 .right)).of_eq fun n => by simp
+#align primrec.unpair Primrec.unpair
+
+theorem list_get?₁ : ∀ l : List α, Primrec l.get?
+  | [] => dom_denumerable.2 zero
+  | a :: l =>
+    dom_denumerable.2 <|
+      (cases1 (encode a).succ <| dom_denumerable.1 <| list_get?₁ l).of_eq fun n => by
+        cases n <;> simp
+#align primrec.list_nth₁ Primrec.list_get?₁
+
+end Primrec
+
+/-- `primrec₂ f` means `f` is a binary primitive recursive function.
+  This is technically unnecessary since we can always curry all
+  the arguments together, but there are enough natural two-arg
+  functions that it is convenient to express this directly. -/
```

Comment:
```suggestion
/-- `Primrec₂ f` means `f` is a binary primitive recursive function.
  This is technically unnecessary since we can always curry all
  the arguments together, but there are enough natural two-arg
  functions that it is convenient to express this directly. -/
```